homeassistant/components/tado/__init__.py

```diff
@@ -38,91 +44,103 @@ def setup(hass, config):
     username = config[DOMAIN][CONF_USERNAME]
     password = config[DOMAIN][CONF_PASSWORD]
 
-    try:
-        tado = Tado(username, password)
-        tado.setDebugging(True)
-    except (RuntimeError, urllib.error.HTTPError):
-        _LOGGER.error("Unable to connect to mytado with username and password")
+    tadoconnector = TadoConnector(hass, username, password)
+    if not tadoconnector.setup():
         return False
 
-    hass.data[DATA_TADO] = TadoDataStore(tado)
+    hass.data[DOMAIN] = tadoconnector
+
+    # Do first update
+    tadoconnector.update()
 
+    # Load components
     for component in TADO_COMPONENTS:
-        load_platform(hass, component, DOMAIN, {}, config)
+        load_platform(
+            hass,
+            component,
+            DOMAIN,
+            {CONF_FALLBACK: config[DOMAIN][CONF_FALLBACK]},
+            config,
+        )
+
+    # Poll for updates in the background
+    hass.helpers.event.track_time_interval(
+        lambda now: tadoconnector.update(), SCAN_INTERVAL
+    )
 
     return True
 
 
-class TadoDataStore:
+class TadoConnector:
     """An object to store the Tado data."""
 
-    def __init__(self, tado):
-        """Initialize Tado data store."""
-        self.tado = tado
-
-        self.sensors = {}
-        self.data = {}
+    def __init__(self, hass, username, password):
+        """Initialize Tado Connector."""
+        self.hass = hass
+        self._username = username
+        self._password = password
+
+        self.tado = None
+        self.zones = None
+        self.devices = None
+        self.data = {
+            "zone": {},
+            "device": {},
+        }
+
+    def setup(self):
+        """Connect to Tado and fetch the zones."""
+        try:
+            self.tado = Tado(self._username, self._password)
+            self.tado.setDebugging(True)
```

Comment:
Can this line raise?

Reply:
Nope, I've moved it out of the try except block.